samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -758,41 +744,50 @@ public void run() {
       sideInputSystemConsumers.register(ssp, startingOffset);
       taskInstanceMetrics.get(this.sspSideInputHandlers.get(ssp).getTaskName()).addOffsetGauge(
           ssp, ScalaJavaUtil.toScalaFunction(() -> this.sspSideInputHandlers.get(ssp).getLastProcessedOffset(ssp)));
+      sideInputTaskMetrics.get(this.sspSideInputHandlers.get(ssp).getTaskName()).addOffsetGauge(
+          ssp, ScalaJavaUtil.toScalaFunction(() -> this.sspSideInputHandlers.get(ssp).getLastProcessedOffset(ssp)));
+    }
 
-      SystemStreamMetadata systemStreamMetadata = streamMetadataCache.getSystemStreamMetadata(ssp.getSystemStream(), false);
-      SystemStreamMetadata.SystemStreamPartitionMetadata sspMetadata =
-          (systemStreamMetadata == null) ? null : systemStreamMetadata.getSystemStreamPartitionMetadata().get(ssp.getPartition());
+    Map<TaskName, TaskSideInputHandler> taskSideInputHandlers = this.sspSideInputHandlers.values().stream()
+        .distinct()
+        .collect(Collectors.toMap(TaskSideInputHandler::getTaskName, Function.identity()));
 
-      // record a copy of the sspMetadata, to later check if its caught up
-      initialSideInputSSPMetadata.put(ssp, sspMetadata);
+    Map<TaskName, RunLoopTask> sideInputTasks = new HashMap<>();
+    this.taskSideInputStoreSSPs.forEach((taskName, storesToSSPs) -> {
+        Set<SystemStreamPartition> taskSSPs = this.taskSideInputStoreSSPs.get(taskName).values().stream()
+            .flatMap(Set::stream)
+            .collect(Collectors.toSet());
 
-      // check if the ssp is caught to upcoming, even at start
-      checkSideInputCaughtUp(ssp, startingOffset, SystemStreamMetadata.OffsetType.UPCOMING, false);
```

Comment:
just confirming: the check is the same now because StorageManagerUtil.getStartingOffset will get the Upcoming offset and from the definition of Upcoming and Newest offset - Upcoming > Newest.  So, although the new checkCaughtUp replacing this method is not explicitly using the OffsetTypes, it is still achieving the same.

Reply:
Yes, the check remains the same. Rather than conditioning the comparator input on upcoming vs newest, we now always check against newest but treat the comparator result slightly differently (i.e. now the startingOffset must be _greater_ than the newest offset when checked initially, rather than before checking equality on the upcoming offset).

Is that clear enough?

Reply:
reverting this to keep the original logic